packages/xchacha20poly1305/xchacha20poly1305.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2019 Kyle Den Hartog
+// MIT License. See LICENSE file for details.
+
+import { AEAD } from "@stablelib/aead";
+import { hchacha } from "@stablelib/xchacha20";
+import { ChaCha20Poly1305 } from "@stablelib/chacha20poly1305";
+import { wipe } from "@stablelib/wipe";
+
+
+export const KEY_LENGTH = 32;
+export const NONCE_LENGTH = 24;
+export const TAG_LENGTH = 16;
+
+/**
+ * XChaCha20-Poly1305 Authenticated Encryption with Associated Data.
+ *
+ * Defined in draft-irtf-cfrg-xchacha.
+ * see https://tools.ietf.org/html/draft-irtf-cfrg-xchacha-01
+ */
+export class XChaCha20Poly1305 implements AEAD {
+  readonly nonceLength = NONCE_LENGTH;
+  readonly tagLength = TAG_LENGTH;
+
+  private _key: Uint8Array;
+
+  /**
+   * Creates a new instance with the given 32-byte key.
+   */
+  constructor(key: Uint8Array) {
+    if (key.length !== KEY_LENGTH) {
+      throw new Error("ChaCha20Poly1305 needs 32-byte key");
+    }
+    // Copy key.
+    this._key = new Uint8Array(key);
+  }
+
+  /**
+   * Encrypts and authenticates plaintext, authenticates associated data,
+   * and returns sealed ciphertext, which includes authentication tag.
+   *
+   * draft-irtf-cfrg-xchacha defines a 24 byte nonce (192 bits) which
+   * then uses the first 16 bytes of the nonce and the secret key with
+   * Hchacha to generate an initial subkey. The last 8 bytes of the nonce
+   * are then prefixed with 4 null bytes and then provided with the subkey
```

Comment:
"4 zero bytes"

Reply:
Fixed for all of the times where it occurred.